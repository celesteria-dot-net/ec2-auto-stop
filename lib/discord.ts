import { format } from 'date-fns';
import { utcToZonedTime } from 'date-fns-tz';
import { Response } from 'node-fetch';
import post from './httpWrapper';

type webhookContents = {
  text: string;
  embed: string;
  color: number;
};

const normalContents: webhookContents = {
  text: '',
  embed: '実行中のインスタンスはありません。',
  color: 1127128,
};

const abnormalContents: webhookContents = {
  text: '@everyone EC2に関して今すぐ確認が必要です！',
  embed: 'インスタンスが実行中です。',
  color: 14177041,
};

const postToDiscord = async (instanceIds: string[]): Promise<Response> => {
  const webhook = process.env.WEBHOOK_EC2;
  const contents = instanceIds.length === 0 ? normalContents : abnormalContents;
  const requestData = {
    username: 'EC2-Auto-Stop',
    avatar_url: 'https://i.gyazo.com/5e6236efcc9c6d6f3bbf6253aa38ea31.png',
    content: `[${format(
      utcToZonedTime(new Date(), 'Asia/Tokyo'),
      'yyyy/MM/dd HH:mm:ss',
    )}]\n${contents.text}`,
    embeds: [
      {
        title: 'EC2 Notifications',
        url:
          'https://us-west-2.console.aws.amazon.com/ec2/v2/home?region=us-west-2#Instances:',
        description: contents.embed,
        color: contents.color,
        fields: [
          {
            name: 'instance_ids',
            value: instanceIds.length === 0 ? '[]' : instanceIds.join(', '),
          },
        ],
        timestamp: new Date().toISOString(),
      },
    ],
  };

  return post(webhook, requestData);
};

export default postToDiscord;
