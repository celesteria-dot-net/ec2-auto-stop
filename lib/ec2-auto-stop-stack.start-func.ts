import { Filter } from '@aws-sdk/client-ec2';
import { startedInstances, noInstancesToRun } from './domains/webhookContents';
import postToDiscord from './util/discord';
import { fetchInstanceIds, startInstances } from './util/ec2';

const tagNames = ['bot_container']

const filters: Filter[] = [
  { Name: 'tag-key', Values: tagNames },
]

// eslint-disable-next-line import/prefer-default-export
export const handler = async (): Promise<{
  isSuccessful: boolean;
  error?: unknown;
}> => {
  try {
    const instanceIds = await fetchInstanceIds(filters);
    if (instanceIds.length !== 0) await startInstances(instanceIds);

    const webhookMessage =
      instanceIds.length === 0 ? noInstancesToRun : startedInstances;
    const response = await postToDiscord(instanceIds, webhookMessage);
    if (response.status !== 204)
      return {
        isSuccessful: false,
        error: `Webhook error: ${response.statusText}`,
      };

    return {
      isSuccessful: true,
    };
  } catch (err) {
    return {
      isSuccessful: false,
      error: err,
    };
  }
};
