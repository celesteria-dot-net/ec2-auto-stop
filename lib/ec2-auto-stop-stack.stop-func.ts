import {
  EC2Client,
  DescribeInstancesCommand,
  StopInstancesCommand,
} from '@aws-sdk/client-ec2';
import fetch from 'node-fetch';

const client = new EC2Client({});

const fetchInstanceIds = async () => {
  const data = await client.send(
    new DescribeInstancesCommand({
      Filters: [
        { Name: 'instance-state-name', Values: ['running', 'pending'] },
      ],
    }),
  );

  return (
    data.Reservations?.flatMap((res) => res.Instances ?? []).flatMap(
      (ins) => ins.InstanceId ?? [],
    ) ?? []
  );
};

const stopInstances = async (instanceIds: string[]) => {
  await client.send(
    new StopInstancesCommand({ InstanceIds: instanceIds, Force: true }),
  );
};

const postToDiscord = async (instanceIds: string[]) => {
  const webhook = process.env.WEBHOOK;
  const normalContent = {
    textContent: '',
    embedContent: 'There are no running instances.',
    color: 1127128,
  };
  const abnormalContent = {
    textContent: '@everyone Check It Now!',
    embedContent: 'Some instances should be stopped!',
    color: 14177041,
  };
  const requestHeader = {
    'User-Agent': 'curl/7.74.0',
    'Content-Type': 'application/json',
  };
  const contents = instanceIds.length === 0 ? normalContent : abnormalContent;
  const requestData = {
    content: contents.textContent,
    embeds: [
      {
        title: 'EC2 Notifications',
        url:
          'https://us-west-2.console.aws.amazon.com/ec2/v2/home?region=us-west-2#Instances:',
        description: contents.embedContent,
        color: contents.color,
        fields: [
          {
            name: 'instance_ids',
            value: instanceIds.length === 0 ? '[]' : instanceIds.join(', '),
          },
        ],
        timestamp: new Date().toISOString(),
      },
    ],
  };

  const response = await fetch(webhook, {
    method: 'post',
    body: JSON.stringify(requestData),
    headers: requestHeader,
  });

  return response;
};

// eslint-disable-next-line import/prefer-default-export
export const handler = async (): Promise<{
  isSuccessful: boolean;
  error?: unknown;
}> => {
  try {
    const instanceIds = await fetchInstanceIds();
    if (instanceIds.length !== 0) await stopInstances(instanceIds);

    const response = await postToDiscord(instanceIds);
    if (response.status !== 204)
      return {
        isSuccessful: false,
        error: `Webhook error: ${response.statusText}`,
      };

    return {
      isSuccessful: true,
    };
  } catch (err) {
    return {
      isSuccessful: false,
      error: err,
    };
  }
};
