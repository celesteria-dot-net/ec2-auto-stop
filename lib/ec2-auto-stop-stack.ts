import * as cdk from '@aws-cdk/core';
import { NodejsFunction } from '@aws-cdk/aws-lambda-nodejs';
import { PolicyStatement } from '@aws-cdk/aws-iam';
import { Rule, Schedule } from '@aws-cdk/aws-events';
import { LambdaFunction } from '@aws-cdk/aws-events-targets';

// eslint-disable-next-line import/prefer-default-export
export class Ec2AutoStopStack extends cdk.Stack {
  constructor(scope: cdk.Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    const policyStatements = [
      new PolicyStatement({
        resources: [
          `arn:aws:ec2:${cdk.Aws.REGION}:${cdk.Aws.ACCOUNT_ID}:instance/*`,
        ],
        actions: ['ec2:StopInstances'],
      }),
      // DescribeInstancesはリソースの指定ができない(ref. https://dev.classmethod.jp/articles/tsnote-iam-resource-level-001/)
      new PolicyStatement({
        resources: ['*'],
        actions: ['ec2:DescribeInstances'],
      }),
    ];

    const stopFunc = new NodejsFunction(this, 'stop-func');
    policyStatements.forEach((policy) => {
      stopFunc.addToRolePolicy(policy);
    });

    const stopFuncTriggers = [
      new Rule(this, 'stop-func-schedule', {
        schedule: Schedule.cron({
          minute: '30,40',
          hour: '14',
        }),
      }),
    ];

    stopFuncTriggers.forEach((rule) =>
      rule.addTarget(new LambdaFunction(stopFunc)),
    );
  }
}
