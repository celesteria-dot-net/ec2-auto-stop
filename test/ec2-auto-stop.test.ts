import { SynthUtils } from '@aws-cdk/assert';
import * as cdk from '@aws-cdk/core';
import * as Ec2AutoStop from '../lib/ec2-auto-stop-stack';

test('Ec2AutoStopStack Snapshot Test', () => {
  const app = new cdk.App();
  const stack = new Ec2AutoStop.Ec2AutoStopStack(app, 'MyTestStack');
  expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
});
