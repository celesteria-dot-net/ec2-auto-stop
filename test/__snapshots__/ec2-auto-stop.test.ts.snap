// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Ec2AutoStopStack Snapshot Test stack 1`] = `
Object {
  "Parameters": Object {
    "AssetParametersda3caf97e9eeeeaaa1dbedb76017b046a53f6a50cfb2378fb7b05c6dec83e519ArtifactHash4697F25C": Object {
      "Description": "Artifact hash for asset \\"da3caf97e9eeeeaaa1dbedb76017b046a53f6a50cfb2378fb7b05c6dec83e519\\"",
      "Type": "String",
    },
    "AssetParametersda3caf97e9eeeeaaa1dbedb76017b046a53f6a50cfb2378fb7b05c6dec83e519S3Bucket395A71AD": Object {
      "Description": "S3 bucket for asset \\"da3caf97e9eeeeaaa1dbedb76017b046a53f6a50cfb2378fb7b05c6dec83e519\\"",
      "Type": "String",
    },
    "AssetParametersda3caf97e9eeeeaaa1dbedb76017b046a53f6a50cfb2378fb7b05c6dec83e519S3VersionKey944084CA": Object {
      "Description": "S3 key for asset version \\"da3caf97e9eeeeaaa1dbedb76017b046a53f6a50cfb2378fb7b05c6dec83e519\\"",
      "Type": "String",
    },
  },
  "Resources": Object {
    "stopfuncFD6C7B5F": Object {
      "DependsOn": Array [
        "stopfuncServiceRoleDefaultPolicyEE15E9C6",
        "stopfuncServiceRole0C10635E",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParametersda3caf97e9eeeeaaa1dbedb76017b046a53f6a50cfb2378fb7b05c6dec83e519S3Bucket395A71AD",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParametersda3caf97e9eeeeaaa1dbedb76017b046a53f6a50cfb2378fb7b05c6dec83e519S3VersionKey944084CA",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParametersda3caf97e9eeeeaaa1dbedb76017b046a53f6a50cfb2378fb7b05c6dec83e519S3VersionKey944084CA",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Environment": Object {
          "Variables": Object {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
          },
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "stopfuncServiceRole0C10635E",
            "Arn",
          ],
        },
        "Runtime": "nodejs12.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "stopfuncServiceRole0C10635E": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "stopfuncServiceRoleDefaultPolicyEE15E9C6": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "ec2:StopInstances",
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:aws:ec2:",
                    Object {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    Object {
                      "Ref": "AWS::AccountId",
                    },
                    ":instance/*",
                  ],
                ],
              },
            },
            Object {
              "Action": "ec2:DescribeInstances",
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "stopfuncServiceRoleDefaultPolicyEE15E9C6",
        "Roles": Array [
          Object {
            "Ref": "stopfuncServiceRole0C10635E",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "stopfuncschedule55A0A613": Object {
      "Properties": Object {
        "ScheduleExpression": "cron(30,40 14 * * ? *)",
        "State": "ENABLED",
        "Targets": Array [
          Object {
            "Arn": Object {
              "Fn::GetAtt": Array [
                "stopfuncFD6C7B5F",
                "Arn",
              ],
            },
            "Id": "Target0",
          },
        ],
      },
      "Type": "AWS::Events::Rule",
    },
    "stopfuncscheduleAllowEventRuleEc2AutoStopStackstopfunc55F42AFB39CCEFA0": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "stopfuncFD6C7B5F",
            "Arn",
          ],
        },
        "Principal": "events.amazonaws.com",
        "SourceArn": Object {
          "Fn::GetAtt": Array [
            "stopfuncschedule55A0A613",
            "Arn",
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
  },
}
`;
